java-checks/src/main/java/org/sonar/java/symexec/SymbolicEvaluator.java

```diff
@@ -370,177 +398,198 @@ public final void visitMemberSelectExpression(MemberSelectExpressionTree tree) {
+        }
+      }
     }
 
-    @Override
-    public void visitBlock(BlockTree tree) {
-      for (StatementTree statement : tree.body()) {
-        currentStates = evaluateStatement(currentStates, statement);
+    private void evaluateUnaryWithSizeEffect(UnaryExpressionTree tree) {
+      for (ExecutionState executionState : currentStates) {
+        Symbol.VariableSymbol symbol = extractVariableSymbol(tree.expression());
+        if (symbol != null) {
+          executionState.assignValue(symbol, createSymbolicInstanceValue());
+        }
       }
     }
 
     @Override
     public void visitBreakStatement(BreakStatementTree tree) {
-      currentStates = PackedStatementStates.instantiateWithBreakStates(currentStates.states);
+      currentStates.breakStates.addAll(currentStates.states);
+      currentStates.states.clear();
+    }
+
+    @Override
+    public void visitCaseGroup(CaseGroupTree tree) {
+      scan(tree.body());
     }
 
     @Override
     public void visitContinueStatement(ContinueStatementTree tree) {
-      currentStates = PackedStatementStates.instantiate();
+      currentStates = new PackedStatementStates(currentStates);
     }
 
     @Override
     public void visitDoWhileStatement(DoWhileStatementTree tree) {
+      PackedStatementStates nextStates = new PackedStatementStates(currentStates);
       Set<Symbol.VariableSymbol> assignedSymbols = extractor.findAssignedVariables(tree);
       invalidateAssignedVariables(assignedSymbols);
-      currentStates = evaluateStatement(currentStates, tree.statement());
-      PackedStatementStates nextStates = PackedStatementStates.instantiate();
-      for (ExecutionState state : currentStates) {
-        if (evaluateExpression(state, tree.condition()) != SymbolicBooleanConstraint.TRUE) {
-          nextStates.addState(state);
+      scan(tree.statement());
+      scan(tree.condition());
+      for (ExecutionState executionState : currentStates) {
+        if (!executionState.popLast().equals(SymbolicValue.BOOLEAN_TRUE)) {
+          nextStates.addState(executionState);
         }
       }
+      nextStates.addStates(currentStates.breakStates);
       currentStates = nextStates;
       invalidateAssignedVariables(assignedSymbols);
     }
 
     @Override
     public void visitExpressionStatement(ExpressionStatementTree tree) {
-      for (ExecutionState state : currentStates) {
-        evaluateExpression(state, tree.expression());
+      scan(tree.expression());
+      for (ExecutionState executionState : currentStates) {
+        executionState.popLast();
       }
     }
 
     @Override
     public void visitForStatement(ForStatementTree tree) {
+      scan(tree.initializer());
+      if (tree.condition() != null) {
+        evaluateForStatement(tree);
+      } else {
+        evaluateForStatementWithoutCondition(tree);
+      }
+    }
+
+    private void evaluateForStatement(ForStatementTree tree) {
+      // FIXME: variables in initializer should not be invalidated.
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1134)